### PR TITLE
Enh: Speed up shapes box selection by using bounding boxes and vectorization

### LINKED ISSUE #8373: [Performance, shapes] Selecting using the drag box is slow when layer has 5000 shapes
### 🐛 Bug Report

If you have a shapes layer with a lot of shapes, selecting shapes using the drag box is slow for both cases:
- if you zoom in and select one
- zoom out and select many

It takes ~15-20 s of blocking. Meanwhile, selecting one by click or all using keybind is instant.
Doesn't matter if path or polygon.

### 💡 Steps to Reproduce

```
import napari
import numpy as np

def create_sample_coords(n_shapes=5000, n_rays=32):
    radius = 500/np.sqrt(n_shapes)
    center = np.random.uniform(0,1000,(n_shapes,2))
    phi = np.linspace(0,2*np.pi,n_rays)
    rays = np.stack([np.sin(phi), np.cos(phi)],axis=1)
    rays = rays.reshape((1,-1,2))
    rays = rays*np.random.uniform(.9,1.1, (n_shapes,n_rays,2))
    center = center.reshape((-1,1,2))
    coords = center + radius*rays
    return coords

coords = create_sample_coords(5000)

viewer = napari.Viewer()
kwargs = dict(shape_type="path",edge_color="coral")

layer = napari.layers.Shapes(coords,**kwargs)

viewer.add_layer(layer)

```

### 💡 Expected Behavior

It should be faster. I can understand some slowness for many shapes in a large selection area, but selecting a small sub-region should be fast.

### 🌎 Environment

napari: 0.6.6.dev2+g6da9831a8
Platform: macOS-14.5-arm64-arm-64bit-Mach-O
System: MacOS 14.5
Python: 3.13.8 | packaged by conda-forge | (main, Oct 13 2025, 14:39:11) [Clang 19.1.7 ]
Qt: 6.10.0
PySide6: 6.10.0
NumPy: 2.3.3
SciPy: 1.16.2
Dask: 2025.9.1
VisPy: 0.15.2
magicgui: 0.10.1
superqt: 0.7.6
in-n-out: 0.2.1
app-model: 0.4.0
psygnal: 0.14.2
npe2: 0.7.9
pydantic: 2.12.1

OpenGL:
- PyOpenGL: 3.1.10
- GL version: 2.1 Metal - 88.1
- MAX_TEXTURE_SIZE: 16384
- GL_MAX_3D_TEXTURE_SIZE: 2048

Screens:
- screen 1: resolution 1680x1050, scale 2.0

Optional:
- numba: 0.62.1
- triangle: 20250106
- napari-plugin-manager: 0.1.6
- bermuda: 0.1.6
- PartSegCore not installed

Experimental Settings:
- Async: True
- Autoswap buffers: False
- Triangulation backend: Fastest available

Settings path:
- /Users/piotrsobolewski/Library/Application Support/napari/napari-dev_a1eb8b76ba95fa16ad06e26097b46b8455dfbf0b/settings.yaml

Launch command:
- /Users/piotrsobolewski/Dev/miniforge3/envs/napari-dev/bin/napari

Plugins:
- napari-skimage: 0.6.0 (67 contributions)
- napari-tiff: 0.1.5.dev12+gc4a37e445 (2 contributions)
- napari: 0.6.6.dev2+g6da9831a8 (88 contributions)
- yanse: 0.1.0.dev0 (4 contributions)
- napari-svg: 0.2.1 (2 contributions)
- napari-console: 0.1.3 (0 contributions)

### 💡 Additional Context

Discovered while looking at https://github.com/napari/napari/issues/8264

### PR DESCRIPTION
**Pull request by psobolewskiPhD**
# References and relevant issues
Closes: https://github.com/napari/napari/issues/8373

# Description

Disclaimer: This PR has pretty extensive contributions from Gemini when it comes to the vectorization changes.

In this PR I began by trying to improve `shapes_in_box` by using bounding boxes to not do the `triangles_intersect_box` for every single shape in the layer. That did help some, but not really enough. 
So based on that I turned to Gemini for help, which led to a series of optimizations by using vectorization:
- `triangles_intersect_box` (only called from `shapes_in_box`) was improved to use a vectorized line intersection, intersecting all the segments with the target at once,  using the new function `vectorized_lines_intersect`. The original `lines_intersect` is no longer used anywhere in the code. For now I left it for reference, having it use the new function.
- `vectorized_lines_intersect` (new function) uses the same approach as the original `lines_intersect`. To vectorize it, it needs a vectorized `orientation` function, which has been added to _accelerate_triangulate_python as `vectorized_orientation`. The original `orientation` isn't used anywhere else, but again I left it for reference.

Much of the vectorization code is from Gemini, but it was generated over an extensive conversation, followed by refactoring on my part to keep the original code and localize things where I thought it made sense. 
Also, with some googling I found the solution to `np.cross` being deprecated which was causing failing tests, so I followed the advice of the numpy devs in the issue, see: https://github.com/numpy/numpy/issues/26620#issuecomment-2150748569

All relevant tests pass for me locally. 
Also the speedup is significant, ~50X. Using the script in the issue (5K) shapes, selecting 90% of them on main takes >25s s, with this PR it's ~0.5 s. For the case of 10K shapes, main is >50s, this PR still <1s. Even doing some sub-selecting out of 150K shapes (script from: https://github.com/napari/napari/issues/7996) is pretty bearable, with just a few second delay. 
Obviously the scaling is O(N) which is not ideal, but probably requires an overall to use something like a spatial index. it may be possible to speed up in `bermuda` or using numba, I'm not sure. But having a numpy version seems worth it, even if it ends up as a stepping stone to those options.

Edit: with the changes from Grzegorz the speedup is now more like 100-150X. For the 10K shape case it's ~0.3 s.

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
For future refactoring @kephale https://github.com/kephale/rstar-python would be relevant!

File: src/napari/layers/shapes/_shapes_utils.py [hunk 1]
```diff
@@ -226,12 +228,12 @@ def triangle_edges_intersect_box(triangles, corners):
...
+            q2 = box[(j + 1) % 4]
+
+            intersects = vectorized_lines_intersect(p1, q1, p2, q2)
+            all_intersects |= intersects
 
-    return np.any(intersects, axis=1)
+    return all_intersects
 
 
 def lines_intersect(p1, q1, p2, q2):
```
Comment (psobolewskiPhD) [inline src/napari/layers/shapes/_shapes_utils.py:239]: This scalar function isn't used elsewhere, so maybe it could be removed with the new function renamed in it's place.
But maybe it can become useful in the future?

File: src/napari/layers/shapes/_shapes_utils.py [hunk 2]
```diff
@@ -253,62 +255,93 @@ def lines_intersect(p1, q1, p2, q2):
...
+    o3 = _triangulate_py.vectorized_orientation(p2, q2, p1)
+    o4 = _triangulate_py.vectorized_orientation(p2, q2, q1)
 
     # Test general case
-    if (o1 != o2) and (o3 != o4):
-        return True
+    intersects = (o1 != o2) & (o3 != o4)
 
-    # Test special cases
+    # Test spacial cases
```
Comment (psobolewskiPhD) [inline src/napari/layers/shapes/_shapes_utils.py]:
```suggestion
    # Test special cases
```

File: src/napari/layers/shapes/_shapes_utils.py [hunk 3]
```diff
@@ -253,62 +255,93 @@ def lines_intersect(p1, q1, p2, q2):
...
-    if o4 == 0 and on_segment(p2, q1, q2):  # noqa: SIM103
-        return True
+    intersects |= (o4 == 0) & on_segment(p2, q1, q2)
+
+    return intersects
 
-    # Doesn't fall into any special cases
-    return False
 
+def on_segment(p: npt.NDArray, q: npt.NDArray, r: npt.NDArray) -> npt.NDArray:
```
Comment (psobolewskiPhD) [inline src/napari/layers/shapes/_shapes_utils.py:319]: This helper wasn't used anywhere else, so I felt it was Ok to just go with the refactor.

File: src/napari/layers/shapes/_accelerated_triangulate_python.py [hunk 4]
```diff
@@ -482,6 +482,60 @@ def orientation(p: np.ndarray, q: np.ndarray, r: np.ndarray) -> np.ndarray:
     return val
 
 
+def vectorized_orientation(
```
Comment (psobolewskiPhD) [inline src/napari/layers/shapes/_accelerated_triangulate_python.py:486]: Expand up to the original scalar `orientation`. It's not used anywhere after the refactor of this PR. But for now I left it, because maybe it could be useful -- it was short so I left it as is.

**Comment by psobolewskiPhD** [comment]
With this PR I can't reproduce the issue in https://github.com/napari/napari/issues/8431
Could be related to the bounding box filtering first?

**Comment by Czaki** [comment]
I changed the code to use `refresh` instead of connecting to the event.

**Comment by psobolewskiPhD** [comment]
> I changed the code to use `refresh` instead of connecting to the event.

@Czaki I don't see the commit, can you make sure you push?

**Comment by Czaki** [comment]
I like idea in this PR, but few values are already cached in ShapeList so do not need to be recalculated. 

Could you review my idea of improvement of this PR https://github.com/psobolewskiPhD/napari/pull/62

**Comment by Czaki** [comment]
> > I changed the code to use `refresh` instead of connecting to the event.
> 
> @Czaki I don't see the commit, can you make sure you push?

Wrong tab. This was comment to #8536

**Comment by psobolewskiPhD** [comment]
@Czaki in my tests your version is another 3-5x faster!

```python
import napari
import numpy as np
import time

def create_sample_coords(n_shapes=5_000, n_rays=32):
    """Creates sample coordinates for shapes."""
    radius = 500 / np.sqrt(n_shapes)
    center = np.random.uniform(0, 1000, (n_shapes, 2))
    phi = np.linspace(0, 2 * np.pi, n_rays)
    rays = np.stack([np.sin(phi), np.cos(phi)], axis=1)
    rays = rays.reshape((1, -1, 2))
    rays = rays * np.random.uniform(0.9, 1.1, (n_shapes, n_rays, 2))
    center = center.reshape((-1, 1, 2))
    coords = center + radius * rays
    return coords

def profile_box_selection(n_shapes):
    """Profiles selecting most shapes with a bounding box."""
    print(f"--- Profiling for {n_shapes} shapes ---")

    coords = create_sample_coords(n_shapes)
    # We don't need the viewer to run the profiling
    layer = napari.layers.Shapes(coords, shape_type='polygon', edge_color='coral')

    # Define a bounding box that encompasses most, but not all, shapes
    # to avoid the "select all" shortcut in the selection logic.
    extent = layer._extent_data
    size = extent[1] - extent[0]
    margin = size * 0.05  # Create a 5% margin on each side
    corners = np.array([extent[0] + margin, extent[1] - margin])

    # Time the box selection process
    start_time = time.time()
    layer.selected_data = layer._data_view.shapes_in_box(corners)
    end_time = time.time()

    elapsed_time = end_time - start_time
    print(
        f"Time to select {len(layer.selected_data)}/{n_shapes} shapes with a box: "
        f"{elapsed_time:.4f} seconds"
    )
    print("-" * (30 + len(str(n_shapes))))


if __name__ == '__main__':
    shape_counts = [500, 1_000, 5_000, 10_000, 150_000]
    for count in shape_counts:
        profile_box_selection(count)
```
main:
```
--- Profiling for 500 shapes ---
Time to select 490/500 shapes with a box: 2.6903 seconds
---------------------------------
--- Profiling for 1000 shapes ---
Time to select 926/1000 shapes with a box: 5.4260 seconds
----------------------------------
--- Profiling for 5000 shapes ---
Time to select 4304/5000 shapes with a box: 27.2272 seconds
----------------------------------
```
This PR:
```
--- Profiling for 500 shapes ---
Time to select 489/500 shapes with a box: 0.0401 seconds
---------------------------------
--- Profiling for 1000 shapes ---
Time to select 921/1000 shapes with a box: 0.0751 seconds
----------------------------------
--- Profiling for 5000 shapes ---
Time to select 4351/5000 shapes with a box: 0.3953 seconds
----------------------------------
--- Profiling for 10000 shapes ---
Time to select 8489/10000 shapes with a box: 0.7886 seconds
-----------------------------------
--- Profiling for 150000 shapes ---
Time to select 123366/150000 shapes with a box: 12.2531 seconds
```

Your branch
```
--- Profiling for 500 shapes ---
Time to select 480/500 shapes with a box: 0.0529 seconds
---------------------------------
--- Profiling for 1000 shapes ---
Time to select 922/1000 shapes with a box: 0.0716 seconds
----------------------------------
--- Profiling for 5000 shapes ---
Time to select 4345/5000 shapes with a box: 0.1929 seconds
----------------------------------
--- Profiling for 10000 shapes ---
Time to select 8502/10000 shapes with a box: 0.2919 seconds
-----------------------------------
--- Profiling for 150000 shapes ---
Time to select 123199/150000 shapes with a box: 1.9789 seconds
```

how do you want to proceed? do you want to just make your own pr? push here?

Edit: wonder how many more places in our codebase there are potential 100-150x speedups waiting discovery...

**Comment by Czaki** [comment]
> how do you want to proceed? do you want to just make your own pr? push here?

I open this PR to this branch. So if you are happy with my changes, just click merge button in your repository. 
I feel that my changes are improvement of your PR. 
The review take me a long time as your PR code looks solid, but even after fixing the bounding box to use cached value I have feling that something could be improved here. And finaly I found that, if we assume that only small number of shapes will intersect bounding boxes then we could do it better. 





> Edit: wonder how many more places in our codebase there are potential 100-150x speedups waiting discovery...

In shapes in multiple places. but until someone is complaining, the fix should wait on remove rendering state from Shapes layer (followup of #8254), as many ofthem depends on iteraction with visible portion of data.

File: src/napari/layers/shapes/_shape_list.py [hunk 5]
```diff
@@ -1816,32 +1816,72 @@ def outlines(
...
+        selection_min = np.min(corners, axis=0)
+        selection_max = np.max(corners, axis=0)
+
+        # If the box encompasses all shapes, just get them directly
+        # layer_min = np.min(shape_mins, axis=0)
+        # layer_max = np.max(shape_maxs, axis=0)
+        # if np.all(selection_min <= layer_min) and np.all(
+        #     selection_max >= layer_max
+        # ):
+        #     return self._visible_shapes_indices.tolist()
```
Comment (psobolewskiPhD) [inline src/napari/layers/shapes/_shape_list.py]: @Czaki is this to be deleted or uncommented?

Comment (Czaki) [reply src/napari/layers/shapes/_shape_list.py]: I forget to do performance tests. It adds two iterations of all set but speeds up single edge cases. My intuition is that it should be removed, but I prefer to do performance testing first. 

But I will have time for this tomorrow evening, so one could do this earlier.

Comment (psobolewskiPhD) [reply src/napari/layers/shapes/_shape_list.py]: I uncommented and tested. Makes the function like 3x faster for the all shapes case, but it's in the ms range and doesn't really affect overall highlighting speed because of all the other overhead.
So we can either leave it, since it is faster or delete it to make the code simpler?

**Comment by psobolewskiPhD** [comment]
Note for me this also closes #8431

**Review by Czaki — APPROVED** [review]
